build/tasks/PinVersions.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+using Microsoft.Build.Framework;
+using RepoTasks.VersionPinning;
+
+namespace RepoTasks
+{
+    public class PinVersions : Microsoft.Build.Utilities.Task
+    {
+        [Required]
+        public string BuildRepositoryRoot { get; set; }
+
+        [Required]
+        public ITaskItem[] PackageSources { get; set; }
+
+        public string GraphSpecsRoot { get; set; }
+
+        public ITaskItem[] AdditionalPackageSources { get; set; }
+
+        public override bool Execute()
+        {
+            if (PackageSources?.Length == 0)
+            {
+                Log.LogError($"Missing PackageSources. At least one item source must be specified.");
+                return false;
+            }
+
+            if (string.IsNullOrEmpty(BuildRepositoryRoot))
+            {
+                Log.LogError("BuildRepositoryRoot argument must be specified.");
+                return false;
+            }
+
+            var graphSpecProvider = !string.IsNullOrEmpty(GraphSpecsRoot)
+                ? new DependencyGraphSpecProvider(GraphSpecsRoot)
+                : DependencyGraphSpecProvider.Default;
+
+            var sources = new List<string>();
+            sources.AddRange(PackageSources.Select(i => i.ItemSpec));
+
+            foreach (var item in AdditionalPackageSources)
+            {
+                var path = item.ItemSpec;
+                if (path.StartsWith("http", StringComparison.OrdinalIgnoreCase))
```

Comment:
This feels weird. It'd be cleaner to construct `PackageSources` in the target with the right set of sources instead.